tests/test_spidermiddleware.py

```diff
@@ -0,0 +1,314 @@
+            yield crawler.crawl()
+        self.assertIn("ValueError exception caught", str(log))
+        self.assertIn("spider_exceptions/ValueError", str(log))
+        # generator return value, 3 items before the error
+        crawler = get_crawler(GeneratorErrorAfterItemsSpider)
+        with LogCapture() as log:
+            yield crawler.crawl()
+        self.assertIn("'item_scraped_count': 3", str(log))
+        self.assertIn("FloatingPointError exception caught", str(log))
+        self.assertIn("spider_exceptions/FloatingPointError", str(log))
+
+    @defer.inlineCallbacks
+    def test_process_spider_exception_from_previous_middleware_input(self):
+        crawler = get_crawler(FromPreviousMiddlewareInputSpider)
+        with LogCapture() as log:
+            yield crawler.crawl()
+        self.assertIn("LookupError exception caught", str(log))
+
+    @defer.inlineCallbacks
+    def test_process_spider_exception_from_previous_middleware_output(self):
+        # non-generator output value
+        crawler = get_crawler(NotAGeneratorFromPreviousMiddlewareOutputSpider)
+        with LogCapture() as log:
+            yield crawler.crawl()
+        self.assertNotIn("UnicodeError exception caught", str(log))
+        # generator output value
+        crawler = get_crawler(GeneratorFromPreviousMiddlewareOutputSpider)
+        with LogCapture() as log:
+            yield crawler.crawl()
+        self.assertIn("'item_scraped_count': 3", str(log))
+        self.assertIn("NameError exception caught", str(log))
+
+    @defer.inlineCallbacks
+    def test_process_spider_exception_do_something(self):
+        crawler = get_crawler(DoSomethingSpider)
+        with LogCapture() as log:
+            yield crawler.crawl()
+        self.assertIn("ImportError exception caught", str(log))
+        self.assertIn("{'value': 10}", str(log))
+        self.assertIn("{'value': 'asdf'}", str(log))
+        self.assertIn("{'value': 'http://example.com'}", str(log))
+        self.assertIn("{'value': 'http://example.org'}", str(log))
+
+    @defer.inlineCallbacks
+    def test_process_spider_exception_invalid_return_value_previous_middleware(self):
+        """ don't catch AssertionError """
+        # on middleware's input
+        crawler1 = get_crawler(InvalidReturnValueFromPreviousMiddlewareInputSpider)
+        with LogCapture() as log1:
+            yield crawler1.crawl()
+        self.assertNotIn("AssertionError exception caught", str(log1))
+        self.assertIn("'spider_exceptions/AssertionError'", str(log1))
+        # on middleware's output
+        crawler2 = get_crawler(InvalidReturnValueFromPreviousMiddlewareOutputSpider)
+        with LogCapture() as log2:
+            yield crawler2.crawl()
+        self.assertNotIn("AssertionError exception caught", str(log2))
+        self.assertIn("'spider_exceptions/AssertionError'", str(log2))
+
+    @defer.inlineCallbacks
+    def test_process_spider_exception_execution_chain(self):
```

Comment:
This test ensures only non already executed `process_spider_output` methods are called when `process_spider_exception` returns an iterable, breaking the exception handling chain :-)